### PR TITLE
[SR-5744] Refactoring action to convert if-let to guard-let and vice versa

### PR DESCRIPTION
**Pull request by Regno**
Implement action to convert from:
```
func foo(idxOpt: Int?) {
  if let idx = idxOpt {
    print(idx)
  }
}
```
to
```
func foo(idxOpt: Int?) {
  guard let idx = idxOpt else {
    return
  }
  print(idx)
}
```
and vice-versa

Resolves new feature request [SR-5744](https://bugs.swift.org/browse/SR-5744).
Refactoring action to convert if-let to guard-let and vice versa

Feedback is very much appreciated!

### CODE REVIEW

**Comment by Regno** [comment]
@swift-ci test

**Comment by theblixguy** [comment]
You can't invoke CI yourself, you need commit access for it.

**Comment by theblixguy** [comment]
cc @akyrtzi

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -2070,6 +2070,211 @@ bool RefactoringActionExpandTernaryExpr::performChange() {
...
+    return false;
+
+  if (Info.ContainedNodes.empty())
+    return false;
+
+  IfStmt *If = nullptr;
+
+  if (Info.ContainedNodes.size() == 1) {
+    if (auto S = Info.ContainedNodes[0].dyn_cast<Stmt*>()) {
+        If = dyn_cast<IfStmt>(S);
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: We should always use 2-spaces indentation. Could you fix other indentation issues as well in this change?

File: lib/IDE/Refactoring.cpp [hunk 2]
```diff
@@ -2070,6 +2070,211 @@ bool RefactoringActionExpandTernaryExpr::performChange() {
...
+  }
+
+  if (!If)
+    return true; // abort
+
+  auto CondList = If->getCond();
+
+  if (CondList.size() != 1)
+    return true; // abort
+
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: We can assume `performChange()` is called only when `isApplicable` returns true. So it seems the above logics are unnecessary.

File: lib/IDE/Refactoring.cpp [hunk 3]
```diff
@@ -2070,6 +2070,211 @@ bool RefactoringActionExpandTernaryExpr::performChange() {
...
+    // Get if-let condition
+    SourceRange range = E.getSourceRange();
+    SourceManager &SM = RangeInfo.RangeContext->getASTContext().SourceMgr;
+    auto CondCharRange = Lexer::getCharSourceRangeFromSourceRange(SM, range);
+
+    auto Body = dyn_cast_or_null<BraceStmt>(If->getThenStmt());
+
+    if (!Body)
+        return true; // abort
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Could you move the `ThenStmt` check to the `isApplicable` function too? so we don't need to check it here and surprising the users by aborting.

File: lib/IDE/Refactoring.cpp [hunk 4]
```diff
@@ -2070,6 +2070,211 @@ bool RefactoringActionExpandTernaryExpr::performChange() {
...
+    llvm::StringRef NewLine = "\n";
+
+    OS << tok::kw_guard << Space;
+    OS << CondCharRange.str().str() << Space;
+    OS << tok::kw_else << Space;
+    OS << tok::l_brace << NewLine;
+
+    // Get if-let else body.
+    auto ElseBody = dyn_cast_or_null<BraceStmt>(If->getElseStmt());
+    if (ElseBody) {
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: These two statements can be merged as `if (auto *ElseBody = dyn_cast_or_null<BraceStmt>(If->getElseStmt())) {} `

File: lib/IDE/Refactoring.cpp [hunk 5]
```diff
@@ -2070,6 +2070,211 @@ bool RefactoringActionExpandTernaryExpr::performChange() {
...
+
+    // Get guard condition
+    auto CondList = Guard->getCond();
+    if (CondList.size() != 1)
+        return true; // abort
+
+    auto E = CondList[0];
+    auto P = E.getPatternOrNull();
+    if (!P)
+        return true; // abort
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Similarly, can we move these condition checkings into `isApplicable` function and only check it there?

File: lib/IDE/Refactoring.cpp [hunk 6]
```diff
@@ -2070,6 +2070,211 @@ bool RefactoringActionExpandTernaryExpr::performChange() {
...
+
+    OS << tok::kw_return << NewLine;
+    OS << tok::r_brace << NewLine;
+    OS << BodyCharRange.str().str();
+
+    // Replace if-let to guard
+    auto ReplaceRange = RangeInfo.ContentRange;
+    EditConsumer.accept(SM, ReplaceRange, DeclBuffer.str());
+
+	return false;
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Indentation issue here and all other places.

**Review by nkcsgexi — COMMENTED** [review]
The implementation looks great overall👍! I have only several nitpicks inline.

File: lib/IDE/Refactoring.cpp (hunk 1 above)
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: Fixed.

File: lib/IDE/Refactoring.cpp (hunk 2 above)
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: Removed.

File: lib/IDE/Refactoring.cpp (hunk 3 above)
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: Moved to the `isApplicable`.

File: lib/IDE/Refactoring.cpp (hunk 4 above)
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: Fixed.

File: lib/IDE/Refactoring.cpp (hunk 5 above)
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: Moved.

File: lib/IDE/Refactoring.cpp (hunk 6 above)
Comment (Regno) [reply lib/IDE/Refactoring.cpp]: Fixed.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please stress test

**Comment by Regno** [comment]
Also I fixed similar issues at this PR: [Refactoring action to convert to computed property](https://github.com/apple/swift/pull/25141).

**Comment by nkcsgexi** [comment]
All tests passed. Congratulations for your first Swift contribution!🚢

**Pull request merged**